### PR TITLE
esp32: support ICMPv6 RIO handling

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

For TE7 test cases, Wi-FI devices are required to configure the route table rules based on the RIO messages sent by the Thread border router. The lwIP stack lacks support for this feature.

#### Change overview

* Add support for ICMPv6 RIO handling for ESP32
* Update ESP-IDF release to v4.4 for route hook support.

#### Testing
How was this tested? (at least one bullet point required)
* Manually built examples with ESP-IDF v4.4 release.
* Paired all-cluster-app and checked RIO handling with a Thread border router.

### CODE REVIEW

**Comment by gjc13** [comment]
@andy31415 @woody-apple We need to update the chip-build-esp32 docker image to switch to the v4.4 release of ESP-IDF.

File: examples/platform/esp32/route_hook/esp_route_table.h
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_table.h:16]: Not openthread here

File: examples/platform/esp32/route_hook/esp_route_table.h
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_table.h:25]: Suggest adding unit suffix

File: examples/platform/esp32/route_hook/esp_route_table.h
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_table.h:54]: Could add details about when this is called, such as on a ND protocol RA being received with RIO optiona

File: examples/platform/esp32/route_hook/esp_route_table.h
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_table.h:57]: Not: typo on `destination` here and next comment block, and `route fond` in next block

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_hook.c:132]: Before reading, make sure icmp_payload_len has enough size

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_hook.c:74]: Before assuming you can cast into payload, you need to check that opt_len would not overrun icmp_payload by checking payload_len. The code here assumes that only size enough for the offset much be checked (2 bytes for next option header) and that the packet is truthful about length of subfields. Please add more size validation before any access to icmp_payload pointer

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_hook.c:85]: Why is byte order not converted here? Is it needed or does lwip store addresses in network order. Cannot remember lwip convention

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
- [ ] Need bounding checking on inner access to raw ICMPv6 payload, since current usage trusts inner lengths without validating against outer lengths

**Comment by andy31415** [comment]
> @andy31415 @woody-apple We need to update the chip-build-esp32 docker image to switch to the v4.4 release of ESP-IDF.

@gjc13 - please create a separate PR updating the docker image and docker version. We can fast-track that and I can push the new image to dockerhub once the PR is merged.

**Comment by cecille** [comment]
I haven't had a chance to review this yet (will get to it today), but we do have a patch up to the main lwip to add RIO support more generally. It's totally up to you how you want to handle this for your platform, but if you wouldn't mind taking a look at the patch so we can get some traction on it, it would be very helpful generally for other folks that also need to implement this on their own platforms.

Patch is here: https://savannah.nongnu.org/patch/?10114

**Comment by gjc13** [comment]
> I haven't had a chance to review this yet (will get to it today), but we do have a patch up to the main lwip to add RIO support more generally. It's totally up to you how you want to handle this for your platform, but if you wouldn't mind taking a look at the patch so we can get some traction on it, it would be very helpful generally for other folks that also need to implement this on their own platforms.
> 
> Patch is here: https://savannah.nongnu.org/patch/?10114

We've discussed the lwIP patch and have concern about directly patching lwIP and diverged version with upstream. Our strategy is to use hook in the Matter repo for the time being and wait for the patch to be merged.

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (gjc13) [reply examples/platform/esp32/route_hook/esp_route_hook.c:85]: IPv6 address doesn't have byte order issue.

Comment (tcarmelveilleux) [reply examples/platform/esp32/route_hook/esp_route_hook.c:87]: There is still an issue where we are copying `prefix_len_bytes` but are not validating before that `(prefix_offset + offset + prefix_len_bytes) <= optlen. That is, if `prefix_len_bytes` is garbage, we are going to read a prefix past the end.

A safer way altogether could be to:
1- Avoid doing all this error-prone buffer bounds arithmetic with many adds
2- Create a `bytes_remaining` helper that starts at `optlen` and gets decremented by each `consuming` and keep checking it's sufficiently big before each write

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (cecille) [inline examples/platform/esp32/route_hook/esp_route_hook.c:77]: Need to add 7 first or this will be one byte low if the prefix len is not a byte multiple.

File: examples/platform/esp32/route_hook/esp_route_table.c
Comment (cecille) [inline examples/platform/esp32/route_hook/esp_route_table.c:120]: I think this might be backwards - longer length overrides preference. https://datatracker.ietf.org/doc/html/rfc4191 S. 3.2

File: examples/all-clusters-app/esp32/README.md
Comment (dhrishi) [inline examples/all-clusters-app/esp32/README.md:49]: If we are moving to v4.4 tag (as you've mentioned above), then it should be `git checkout v4.4` ? (at all places)

File: examples/all-clusters-app/esp32/sdkconfig.defaults
Comment (dhrishi) [inline examples/all-clusters-app/esp32/sdkconfig.defaults:50]: Should this also be added in `sdkconfig_c3devkit.defaults`, `sdkconfig_m5stack.defaults` and any relevant sdkconfigs?

File: examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp
Comment (dhrishi) [inline examples/all-clusters-app/esp32/main/DeviceCallbacks.cpp:89]: Error check? (at all places)

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (dhrishi) [inline examples/platform/esp32/route_hook/esp_route_hook.c:36]: Generally typedefs are suffixed with `_t`. So `esp_route_hook_t` ?

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (dhrishi) [inline examples/platform/esp32/route_hook/esp_route_hook.c:177]: If all the members of hook are not set by you, then better use `calloc` or explicitly memset hook?

**Comment by dhrishi** [comment]
@gjc13 I read from the testing section, but just to be sure, no other changes are required for ESP32 CHIP code to work with IDF v4.4?

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_hook.c:83]: Does ESP32 allow unaligned access? This is indexing byte-wise into a buffer. It is very likely that `icmp_payload` pointer could be a non-word-aligned location, especially if the prefix != 64 bits long. Will this cause an access exception on Xtensa? It for sure would cause an access exception on some (not all) Cortex.

**Review by tcarmelveilleux — COMMENTED** [review]
Most comments addressed.

Remaining:

- [ ] Validate safety of type-punning to `rio_header_t`

File: examples/all-clusters-app/esp32/README.md
Comment (gjc13) [reply examples/all-clusters-app/esp32/README.md:49]: v4.4 tag has not been created yet. We have only the branch.

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (gjc13) [reply examples/platform/esp32/route_hook/esp_route_hook.c:177]: All the hook members are set in the following lines so it's fine to use `malloc` here.

**Comment by gjc13** [comment]
> @gjc13 I read from the testing section, but just to be sure, no other changes are required for ESP32 CHIP code to work with IDF v4.4?

v4.4 adds the lwIP route hook.

**Comment by dhrishi** [comment]
LGTM

**Review by woody-apple — APPROVED** [review]

**Review by dhrishi — APPROVED** [review]

**Comment by gjc13** [comment]
@tcarmelveilleux PTAL

File: examples/platform/esp32/route_hook/esp_route_hook.c
Comment (tcarmelveilleux) [inline examples/platform/esp32/route_hook/esp_route_hook.c:85]: Skip prefix length > 128

**Review by tcarmelveilleux — APPROVED** [review]

**Review by chshu — APPROVED** [review]

**Pull request merged**
